state-sync/storage-service/types/src/lib.rs

```diff
@@ -379,3 +546,166 @@ where
+        LedgerInfoWithSignatures::new(
+            LedgerInfo::new(
+                BlockInfo::new(0, 0, HashValue::zero(), HashValue::zero(), version, 0, None),
+                HashValue::zero(),
+            ),
+            BTreeMap::new(),
+        )
+    }
+
+    fn range(lowest: u64, highest: u64) -> CompleteDataRange<u64> {
+        CompleteDataRange::new(lowest, highest).unwrap()
+    }
+
+    fn get_epochs_request(start: Epoch, end: Epoch) -> StorageServiceRequest {
+        StorageServiceRequest::GetEpochEndingLedgerInfos(EpochEndingLedgerInfoRequest {
+            start_epoch: start,
+            expected_end_epoch: end,
+        })
+    }
+
+    fn get_txns_request(proof: Version, start: Version, end: Version) -> StorageServiceRequest {
+        StorageServiceRequest::GetTransactionsWithProof(TransactionsWithProofRequest {
+            proof_version: proof,
+            start_version: start,
+            end_version: end,
+            include_events: true,
+        })
+    }
+
+    #[test]
+    fn test_complete_data_range() {
+        // good ranges
+        assert_ok!(CompleteDataRange::new(0, 0));
+        assert_ok!(CompleteDataRange::new(10, 10));
+        assert_ok!(CompleteDataRange::new(10, 20));
+        assert_ok!(CompleteDataRange::new(u64::MAX, u64::MAX));
+
+        // degenerate ranges
+        assert_err!(CompleteDataRange::new(1, 0));
+        assert_err!(CompleteDataRange::new(20, 10));
+        assert_err!(CompleteDataRange::new(u64::MAX, 0));
+        assert_err!(CompleteDataRange::new(u64::MAX, 1));
+
+        // range length overflow edge case
+        assert_ok!(CompleteDataRange::new(1, u64::MAX));
+        assert_ok!(CompleteDataRange::new(0, u64::MAX - 1));
+        assert_err!(CompleteDataRange::new(0, u64::MAX));
+    }
+
+    #[test]
+    fn test_data_summary_can_service_epochs_request() {
+        let summary = DataSummary {
+            epoch_ending_ledger_infos: Some(range(100, 200)),
+            ..Default::default()
+        };
+
+        // in range, can service
+
+        assert!(summary.can_service(&get_epochs_request(100, 200)));
+        assert!(summary.can_service(&get_epochs_request(125, 175)));
+        assert!(summary.can_service(&get_epochs_request(100, 100)));
+        assert!(summary.can_service(&get_epochs_request(150, 150)));
+        assert!(summary.can_service(&get_epochs_request(200, 200)));
+
+        // out of range, can't service
+
+        assert!(!summary.can_service(&get_epochs_request(99, 200)));
+        assert!(!summary.can_service(&get_epochs_request(100, 201)));
+        assert!(!summary.can_service(&get_epochs_request(50, 250)));
+        assert!(!summary.can_service(&get_epochs_request(50, 150)));
+        assert!(!summary.can_service(&get_epochs_request(150, 250)));
+
+        // degenerate range, can't service
+
+        assert!(!summary.can_service(&get_epochs_request(150, 149)));
+    }
+
+    #[test]
+    fn test_data_summary_can_service_txns_request() {
```

Comment:
Future nit: for completeness can we add the two missing test cases for txn outputs and accounts?